pgjdbc/src/main/java/org/postgresql/jdbc/PgConnection.java

```diff
@@ -378,10 +384,170 @@ public CallableStatement prepareCall(String sql) throws SQLException {
+   *
+   * @see  #getTypeMapInvertedDirect(java.lang.Class)
+   */
+  private static boolean addInverted(Map<Class<?>, Set<String>> invertedMap, Class<?> clazz, String type) {
+    Set<String> types = invertedMap.get(clazz);
+    if (types == null) {
+      types = new HashSet<String>();
+      invertedMap.put(clazz, types);
+    }
+    boolean added = types.add(type);
+    if (added) {
+      LOGGER.log(Level.FINER, "Added: {0} -> {1}", new Object[] {clazz.getName(), type});
+    } else {
+      LOGGER.log(Level.FINEST, "Not added: {0} -> {1}", new Object[] {clazz.getName(), type});
+    }
+    return added;
+  }
+
+  @Override
+  public Set<String> getTypeMapInvertedDirect(Class<?> clazz) {
+    Set<String> types;
+    synchronized (typemapInvertedDirect) {
+      Map<String, Class<?>> source = typemap;
+      if (source != typemapInvertedDirectSource) {
+        // Recreate the inverted map when the typemap has been changed
+        typemapInvertedDirect.clear();
+        for (Map.Entry<String, Class<?>> entry : source.entrySet()) {
+          addInverted(typemapInvertedDirect, entry.getValue(), entry.getKey());
+        }
+        // Make each type set unmodifiable
+        makeValuesUnmodifiable(typemapInvertedDirect);
+        // Update source to know when to rebuild
+        typemapInvertedDirectSource = source;
+        LOGGER.log(Level.FINE, "  typemapInvertedDirect = {0}", typemapInvertedDirect);
+      }
+      types = typemapInvertedDirect.get(clazz);
+    }
+    if (types == null) {
+      return Collections.emptySet();
+    } else {
+      return types;
+    }
+  }
+
+  /**
+   * The set of all known types mapped to each class - used for type inference.
+   */
+  private final Map<Class<?>, Set<String>> typemapInvertedInherited = new HashMap<Class<?>, Set<String>>();
+
+  /**
+   * The {@link #typemap} present at the time {@link #typemapInvertedInherited} was last
+   * rebuilt.
+   *
+   * @see  #getTypeMapInvertedInherited(java.lang.Class)
+   */
+  private Map<String, Class<?>> typemapInvertedInheritedSource = typemap;
+
+  /**
+   * Recursively adds the class, all super classes, all interfaces,
+   * and all extended interfaces.
+   *
+   * @param invertedMap the map to add to
+   * @param current the class to add
+   * @param type the type that maps to this class
+   *
+   * @see  #addInverted(java.util.Map, java.lang.Class, java.lang.String)
+   * @see  #getTypeMapInvertedInherited(java.lang.Class)
+   */
+  private static void addAllInverted(Map<Class<?>, Set<String>> invertedMap, Class<?> current, String type) {
+    do {
+      if (addInverted(invertedMap, current, type)) {
+        for (Class<?> iface : current.getInterfaces()) {
+          addAllInverted(invertedMap, iface, type);
+        }
+        // Class.getSuperclass() returns null for interfaces, but interfaces can be cast to Object
+        // For sake of completeness, add Object for interfaces, too
```

Comment:
should Object be ignored entirely?

Reply:
The inverted maps now skip Object.  However, a custom type could still be directly mapped to Object:

https://github.com/pgjdbc/pgjdbc/pull/1378/commits/d1a0e9e0c055b0f9e4733ca0c3effc249825181b